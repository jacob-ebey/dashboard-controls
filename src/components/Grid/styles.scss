@import '../colors.scss';
.grid {
  &>[data-row] {
    display: flex;
    border-bottom: 1px solid $color-gray;
    padding: 5px 0;
    &[data-header] {
      &>[data-col] {
        font-weight: 800;
      }
    }
    &:not([data-header]) {
      &:hover {
        background-color: $color-grayLightest;
      }
      &:active {
        background-color: $color-grayLight;
      }
    }
    &>[data-col] {
      padding: 5px;
    }
  }
}