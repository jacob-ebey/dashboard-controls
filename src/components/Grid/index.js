import React from 'react';
import PropTypes from 'prop-types';

import styles from './styles.scss';

export default class Component extends React.Component {
  static propTypes = {
    columns: PropTypes.arrayOf(PropTypes.shape({
      key: PropTypes.oneOfType([
        PropTypes.number,
        PropTypes.string,
      ]).isRequired,
      property: PropTypes.string.isRequired,
      label: PropTypes.string,
    })).isRequired,
    rowKeyProperty: PropTypes.string.isRequired,

    items: PropTypes.array,
    onRowClicked: PropTypes.func,
  };

  render() {
    const { className, items, columns, rowKeyProperty, children, ...rest } = this.props;

    const width = 100 / columns.length;

    return (
      <div
        className={`${styles.grid} ${className || ''}`}
        {...rest}
      >
        <div data-row data-header>
        {
          columns.map(col => (
            <div key={col.key} data-col style={{ width: `${width}%` }}>
              {col.label}
            </div>
          ))
        }
        </div>
        {
          items && items.map(item => (
            <div key={item[rowKeyProperty]} data-row onClick={this._handleRowClicked(item)}>
              {
                columns.map(col => (
                  <div key={col.key} data-col style={{ width: `${width}%` }}>
                    {item[col.property]}
                  </div>
                ))
              }
            </div>
          ))
        }
      </div>
    );
  }

  _handleRowClicked = item => () => {
    const { onRowClicked } = this.props;
    const selection = window.getSelection();

    if (!selection.toString()) {
      if (onRowClicked) {
        onRowClicked(item);
      }
    }
  }
}
