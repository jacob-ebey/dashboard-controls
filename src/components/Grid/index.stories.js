import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import Grid from './index';

const createProps = () => ({
  columns: [
    { key: 'name', label: 'Name', property: 'name' },
    { key: 'value', label: 'Value', property: 'value' },
  ],
  rowKeyProperty: 'value',
  items: [
    { name: 'Name 1', value: 'Value 1' },
    { name: 'Name 2', value: 'Value 2' },
  ],
  onRowClicked: action('Grid: rowClicked'),
});

storiesOf('Grid', module)
  .add('default', () => (
    <Grid {...createProps()}/>
  ))
  .add('text-align default', () => (
    <div style={{ textAlign: 'left' }}>
      <Grid {...createProps()}/>
    </div>
  ));
