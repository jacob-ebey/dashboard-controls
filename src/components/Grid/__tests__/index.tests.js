import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import Grid from '../index';

Enzyme.configure({ adapter: new Adapter() });

window.getSelection = () => ({
  toString: () => undefined,
});

describe('components/Grid - index.js', () => {
  const createProps = () => ({
    columns: [
      { key: 'name', label: 'Name', property: 'name' },
      { key: 'value', label: 'Value', property: 'value' },
    ],
    rowKeyProperty: 'value',
    items: [
      { name: 'Name 1', value: 'Value 1' },
      { name: 'Name 2', value: 'Value 2' },
    ],
    onRowClicked: jest.fn(),
  });

  test('Renders without error', () => {
    const props = createProps();

    const wrapper = shallow(<Grid {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  test('Raises onRowClicked', () => {
    const props = createProps();

    const wrapper = shallow(<Grid {...props} />);
    const dataRow = wrapper
      .find('[data-row]')
      .not('[data-header]')
      .first();

    const onRowClick = dataRow.prop('onClick');

    onRowClick();

    expect(props.onRowClicked).toHaveBeenCalledTimes(1);
    expect(props.onRowClicked).toHaveBeenCalledWith(props.items[0]);
  });
});