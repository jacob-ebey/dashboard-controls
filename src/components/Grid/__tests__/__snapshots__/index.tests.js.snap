// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Grid - index.js Renders without error 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Component
    columns={
      Array [
        Object {
          "key": "name",
          "label": "Name",
          "property": "name",
        },
        Object {
          "key": "value",
          "label": "Value",
          "property": "value",
        },
      ]
    }
    items={
      Array [
        Object {
          "name": "Name 1",
          "value": "Value 1",
        },
        Object {
          "name": "Name 2",
          "value": "Value 2",
        },
      ]
    }
    onRowClicked={[Function]}
    rowKeyProperty="value"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": null,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          data-header={true}
          data-row={true}
        >
          <div
            data-col={true}
            style={
              Object {
                "width": "50%",
              }
            }
          >
            Name
          </div>
          <div
            data-col={true}
            style={
              Object {
                "width": "50%",
              }
            }
          >
            Value
          </div>
        </div>,
        Array [
          <div
            data-row={true}
            onClick={[Function]}
          >
            <div
              data-col={true}
              style={
                Object {
                  "width": "50%",
                }
              }
            >
              Name 1
            </div>
            <div
              data-col={true}
              style={
                Object {
                  "width": "50%",
                }
              }
            >
              Value 1
            </div>
          </div>,
          <div
            data-row={true}
            onClick={[Function]}
          >
            <div
              data-col={true}
              style={
                Object {
                  "width": "50%",
                }
              }
            >
              Name 2
            </div>
            <div
              data-col={true}
              style={
                Object {
                  "width": "50%",
                }
              }
            >
              Value 2
            </div>
          </div>,
        ],
      ],
      "className": "grid ",
      "onRowClicked": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": null,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              data-col={true}
              style={
                Object {
                  "width": "50%",
                }
              }
            >
              Name
            </div>,
            <div
              data-col={true}
              style={
                Object {
                  "width": "50%",
                }
              }
            >
              Value
            </div>,
          ],
          "data-header": true,
          "data-row": true,
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "name",
            "nodeType": "host",
            "props": Object {
              "children": "Name",
              "data-col": true,
              "style": Object {
                "width": "50%",
              },
            },
            "ref": null,
            "rendered": "Name",
            "type": "div",
          },
          Object {
            "instance": null,
            "key": "value",
            "nodeType": "host",
            "props": Object {
              "children": "Value",
              "data-col": true,
              "style": Object {
                "width": "50%",
              },
            },
            "ref": null,
            "rendered": "Value",
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "Value 1",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              data-col={true}
              style={
                Object {
                  "width": "50%",
                }
              }
            >
              Name 1
            </div>,
            <div
              data-col={true}
              style={
                Object {
                  "width": "50%",
                }
              }
            >
              Value 1
            </div>,
          ],
          "data-row": true,
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "name",
            "nodeType": "host",
            "props": Object {
              "children": "Name 1",
              "data-col": true,
              "style": Object {
                "width": "50%",
              },
            },
            "ref": null,
            "rendered": "Name 1",
            "type": "div",
          },
          Object {
            "instance": null,
            "key": "value",
            "nodeType": "host",
            "props": Object {
              "children": "Value 1",
              "data-col": true,
              "style": Object {
                "width": "50%",
              },
            },
            "ref": null,
            "rendered": "Value 1",
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": "Value 2",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              data-col={true}
              style={
                Object {
                  "width": "50%",
                }
              }
            >
              Name 2
            </div>,
            <div
              data-col={true}
              style={
                Object {
                  "width": "50%",
                }
              }
            >
              Value 2
            </div>,
          ],
          "data-row": true,
          "onClick": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "name",
            "nodeType": "host",
            "props": Object {
              "children": "Name 2",
              "data-col": true,
              "style": Object {
                "width": "50%",
              },
            },
            "ref": null,
            "rendered": "Name 2",
            "type": "div",
          },
          Object {
            "instance": null,
            "key": "value",
            "nodeType": "host",
            "props": Object {
              "children": "Value 2",
              "data-col": true,
              "style": Object {
                "width": "50%",
              },
            },
            "ref": null,
            "rendered": "Value 2",
            "type": "div",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": null,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            data-header={true}
            data-row={true}
          >
            <div
              data-col={true}
              style={
                Object {
                  "width": "50%",
                }
              }
            >
              Name
            </div>
            <div
              data-col={true}
              style={
                Object {
                  "width": "50%",
                }
              }
            >
              Value
            </div>
          </div>,
          Array [
            <div
              data-row={true}
              onClick={[Function]}
            >
              <div
                data-col={true}
                style={
                  Object {
                    "width": "50%",
                  }
                }
              >
                Name 1
              </div>
              <div
                data-col={true}
                style={
                  Object {
                    "width": "50%",
                  }
                }
              >
                Value 1
              </div>
            </div>,
            <div
              data-row={true}
              onClick={[Function]}
            >
              <div
                data-col={true}
                style={
                  Object {
                    "width": "50%",
                  }
                }
              >
                Name 2
              </div>
              <div
                data-col={true}
                style={
                  Object {
                    "width": "50%",
                  }
                }
              >
                Value 2
              </div>
            </div>,
          ],
        ],
        "className": "grid ",
        "onRowClicked": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": null,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                data-col={true}
                style={
                  Object {
                    "width": "50%",
                  }
                }
              >
                Name
              </div>,
              <div
                data-col={true}
                style={
                  Object {
                    "width": "50%",
                  }
                }
              >
                Value
              </div>,
            ],
            "data-header": true,
            "data-row": true,
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "name",
              "nodeType": "host",
              "props": Object {
                "children": "Name",
                "data-col": true,
                "style": Object {
                  "width": "50%",
                },
              },
              "ref": null,
              "rendered": "Name",
              "type": "div",
            },
            Object {
              "instance": null,
              "key": "value",
              "nodeType": "host",
              "props": Object {
                "children": "Value",
                "data-col": true,
                "style": Object {
                  "width": "50%",
                },
              },
              "ref": null,
              "rendered": "Value",
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "Value 1",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                data-col={true}
                style={
                  Object {
                    "width": "50%",
                  }
                }
              >
                Name 1
              </div>,
              <div
                data-col={true}
                style={
                  Object {
                    "width": "50%",
                  }
                }
              >
                Value 1
              </div>,
            ],
            "data-row": true,
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "name",
              "nodeType": "host",
              "props": Object {
                "children": "Name 1",
                "data-col": true,
                "style": Object {
                  "width": "50%",
                },
              },
              "ref": null,
              "rendered": "Name 1",
              "type": "div",
            },
            Object {
              "instance": null,
              "key": "value",
              "nodeType": "host",
              "props": Object {
                "children": "Value 1",
                "data-col": true,
                "style": Object {
                  "width": "50%",
                },
              },
              "ref": null,
              "rendered": "Value 1",
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": "Value 2",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                data-col={true}
                style={
                  Object {
                    "width": "50%",
                  }
                }
              >
                Name 2
              </div>,
              <div
                data-col={true}
                style={
                  Object {
                    "width": "50%",
                  }
                }
              >
                Value 2
              </div>,
            ],
            "data-row": true,
            "onClick": [Function],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "name",
              "nodeType": "host",
              "props": Object {
                "children": "Name 2",
                "data-col": true,
                "style": Object {
                  "width": "50%",
                },
              },
              "ref": null,
              "rendered": "Name 2",
              "type": "div",
            },
            Object {
              "instance": null,
              "key": "value",
              "nodeType": "host",
              "props": Object {
                "children": "Value 2",
                "data-col": true,
                "style": Object {
                  "width": "50%",
                },
              },
              "ref": null,
              "rendered": "Value 2",
              "type": "div",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
